src/oss-defog/DefoggerTool.cs

```diff
@@ -97,6 +222,113 @@ public EncodedString(EncodedStringType Type, string Filename, string EncodedText
             }
         }
 
+        /// <summary>
+        ///     The encoded string and type.
+        /// </summary>
+        public class EncodedBlob
+        {
+            public string Filename;
+            public string EncodedText;
+            public string DecodedText;
+
+            public EncodedBlob(string Filename, string EncodedText, string DecodedText)
+            {
+                this.Filename = Filename;
+                this.EncodedText = EncodedText;
+                this.DecodedText = DecodedText;
+            }
+        }
+
+        /// <summary>
+        ///     The encoded archive and type.
+        /// </summary>
+        public class EncodedArchive
+        {
+            public ArchiveFileType Type;
+            public string Filename;
+            public string EncodedText;
+            public Stream DecodedArchive;
+
+            public EncodedArchive(ArchiveFileType Type, string Filename, string EncodedText, Stream DecodedArchive)
+            {
+                this.Type = Type;
+                this.Filename = Filename;
+                this.EncodedText = EncodedText;
+                this.DecodedArchive = DecodedArchive;
+            }
+        }
+
+        /// <summary>
+        ///     The encoded binary and type.
+        /// </summary>
+        public class EncodedBinary
+        {
+            public EXECUTABLE_TYPE Type;
+            public string Filename;
+            public string EncodedText;
+            public Stream DecodedBinary;
+
+            public EncodedBinary(EXECUTABLE_TYPE Type, string Filename, string EncodedText, Stream DecodedBinary)
+            {
+                this.Type = Type;
+                this.Filename = Filename;
+                this.EncodedText = EncodedText;
+                this.DecodedBinary = DecodedBinary;
+            }
+        }
+
+        public static byte[] ReadToEnd(System.IO.Stream stream)
+        {
+            long originalPosition = 0;
+
+            if (stream.CanSeek)
+            {
+                originalPosition = stream.Position;
+                stream.Position = 0;
+            }
+
+            try
+            {
+                byte[] readBuffer = new byte[4096];
+
+                int totalBytesRead = 0;
+                int bytesRead;
+
+                while ((bytesRead = stream.Read(readBuffer, totalBytesRead, readBuffer.Length - totalBytesRead)) > 0)
+                {
+                    totalBytesRead += bytesRead;
+
+                    if (totalBytesRead == readBuffer.Length)
+                    {
+                        int nextByte = stream.ReadByte();
+                        if (nextByte != -1)
+                        {
+                            byte[] temp = new byte[readBuffer.Length * 2];
+                            Buffer.BlockCopy(readBuffer, 0, temp, 0, readBuffer.Length);
```

Comment:
What I understand here is that you are expanding the readBuffer based on the contents of the stream. Is there a way to store the length of the stream in the respective classes? That way, you can do just one allocation and just keep adding bytes to it without using the temp array.

Reply:
We just have Stream which doesn't have a length property. You need to read a stream to its end to find its length.  If this were a string stream we could just use a StreamReader.ReadToEnd(), but its a binary stream so there's no convenient equivalent class I'm aware of.  If there's a more clever way that doesn't use a temp array that's fine but I don't think you can do it with one allocation.

If I rewrote some stuff I could use something like a MemoryStream instead of Stream possibly.

Reply:
hmmm...yea I suspected that might be the reason, feel free decide what to do based on if it could significantly improve performance.

Reply:
I rewrote this entirely without any temp allocations.  Take another look.

Reply:
This looks awesome and much cleaner! :)

Reply:
I looked closer at this and I don't think I need this at all.  Removed the whole thing.